projects/ngx-morse/src/lib/ngx-morse.service.spec.ts

```diff
@@ -1,12 +1,83 @@
 import { TestBed } from '@angular/core/testing';
+import InvalidInputError from './exceptions/invalid-input';
 
 import { NgxMorseService } from './ngx-morse.service';
 
 describe('NgxMorseService', () => {
-  beforeEach(() => TestBed.configureTestingModule({}));
 
-  it('should be created', () => {
-    const service: NgxMorseService = TestBed.get(NgxMorseService);
-    expect(service).toBeTruthy();
+  let ngxMorseService: NgxMorseService;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [
+        NgxMorseService,
+      ]
+    });
+
+    ngxMorseService = TestBed.get(NgxMorseService);
+  });
+
+  it('encoding method should exist', () => {
+    expect(ngxMorseService.encode).toBeTruthy();
+  })
+
+  it('decoding method should exist', () => {
+    expect(ngxMorseService.decode).toBeTruthy();
+  })
+
+  // Encode tests
+
+  it('encode: should skip encoding an empty string', () => {
+    const emptyString: string = '';
+    expect(ngxMorseService.encode(emptyString)).toBe(emptyString);
+  });
+
+  it('encode: should encode same irrespective of lowercase or uppercase', () => {
+    const upperCaseMessage: string = 'CASE';
+    const lowerCaseMessage = upperCaseMessage.toLowerCase();
+    const expectedResult = '-.-. .- ... . ';
+
+    expect(ngxMorseService.encode(upperCaseMessage)).toBe(expectedResult);
+    expect(ngxMorseService.encode(lowerCaseMessage)).toBe(expectedResult);
+  });
+
+  it('encode: should encode message made up of valid input', () => {
+    const message: string = 'Keep the dream alive';
+    expect(ngxMorseService.encode(message)).toBe('-.- . . .--. / - .... . / -.. .-. . .- -- / .- .-.. .. ...- . ');
+  });
+
+  it('encode: throw InvalidInputError when encounters invalid input', () => {
+    const message: string = 'Keep the dream alive\t\t\t';
+    expect(() => {
+      // https://stackoverflow.com/a/4144803/5056792
+      ngxMorseService.encode(message);
+    }).toThrow(new InvalidInputError(InvalidInputError.INVALID_INPUT));
+  });
+
+  // Decode tests
+
+  it('decode: should skip decoding an empty string', () => {
```

Comment:
Currently ngxMorseService.decode(< empty string >) returns undefined. We can also throw an invalid input error in that case. what are your thoughts on this @monkeyscript ?

Reply:
Yeah. It's a valid scenario.